blocks/cards/cards.js

```diff
@@ -1,15 +1,50 @@
 import { createOptimizedPicture } from '../../scripts/lib-franklin.js';
 
+const isValidUrl = (urlString) => {
+  try {
+    return Boolean(new URL(urlString));
+  } catch (e) {
+    return false;
+  }
+};
+
 export default function decorate(block) {
```

Comment:
To make the code more readable, we could:
- put logic for processing rows and divs into separate functions
- combine conditions
- replace forEach with a functional version

```
export default function decorate(block) {
  const ul = document.createElement('ul');

  const processRow = (row) => {
    const li = document.createElement('li');
    let cardLink;

    const processDiv = (div) => {
      const { children } = div;
      const picture = div.querySelector('picture');

      if (children.length === 1 && picture) {
        div.className = 'cards-card-image';
      } else {
        div.className = children.length ? 'cards-card-body' : 'cards-card-body empty';
        const action = div.querySelector('p > a');

        if (action) {
          const actionBlock = action.parentElement;
          actionBlock.className = isValidUrl(action.text) ? 'callout hidden' : 'callout';
          div.classList.toggle('callout-hidden', actionBlock.classList.contains('hidden'));
          actionBlock.innerHTML = action.innerHTML;

          cardLink = document.createElement('a');
          cardLink.href = action.href;
          cardLink.innerHTML = row.innerHTML;
        }
      }
    };

    [...row.children].forEach(processDiv);

    if (cardLink) {
      li.appendChild(cardLink);
    } else {
      li.innerHTML = row.innerHTML;
    }
    return li;
  };

  const processImg = (img) => {
    const picture = img.closest('picture');
    picture.replaceWith(createOptimizedPicture(img.src, img.alt, false, null, null, [{ width: '600' }]));
  };

  [...block.children].forEach((row) => {
    ul.append(processRow(row));
  });

  ul.querySelectorAll('img').forEach(processImg);

  block.textContent = '';
  block.append(ul);
}
```